blockchain/emulator.go

```diff
@@ -0,0 +1,112 @@
+	flowgo "github.com/onflow/flow-go/model/flow"
+
+	"github.com/onflow/flow-emulator/types"
+)
+
+type CoverageReportCapable interface {
+	CoverageReport() *runtime.CoverageReport
+	SetCoverageReport(coverageReport *runtime.CoverageReport)
+	ResetCoverageReport()
+}
+
+type DebuggingCapable interface {
+	SetDebugger(*interpreter.Debugger)
+	EndDebugging()
+	// Deprecated: Needed for the debugger right now, do NOT use for other purposes.
+	// TODO: refactor
+	GetAccountUnsafe(address flowgo.Address) (*flowgo.Account, error)
+}
+
+type SnapshotCapable interface {
+	Snapshots() ([]string, error)
+	CreateSnapshot(name string) error
+	LoadSnapshot(name string) error
+}
+
+type RollbackCapable interface {
+	RollbackToBlockHeight(height uint64) error
+}
+
+type AccessProvider interface {
+	Ping() error
+	GetNetworkParameters() access.NetworkParameters
+
+	GetLatestBlock() (*flowgo.Block, error)
+	GetBlockByID(id flowgo.Identifier) (*flowgo.Block, error)
+	GetBlockByHeight(height uint64) (*flowgo.Block, error)
+
+	GetCollectionByID(colID flowgo.Identifier) (*flowgo.LightCollection, error)
+
+	GetTransaction(txID flowgo.Identifier) (*flowgo.TransactionBody, error)
+	GetTransactionResult(txID flowgo.Identifier) (*access.TransactionResult, error)
+	GetTransactionResultByIndex(blockID flowgo.Identifier, index uint32) (*access.TransactionResult, error)
+	GetTransactionsByBlockID(blockID flowgo.Identifier) ([]*flowgo.TransactionBody, error)
+	GetTransactionResultsByBlockID(blockID flowgo.Identifier) ([]*access.TransactionResult, error)
+
+	GetAccount(address flowgo.Address) (*flowgo.Account, error)
+	GetAccountAtBlockHeight(address flowgo.Address, blockHeight uint64) (*flowgo.Account, error)
+	GetAccountByIndex(uint) (*flowgo.Account, error)
+
+	GetEventsByHeight(blockHeight uint64, eventType string) ([]flowgo.Event, error)
+	GetEventsForBlockIDs(eventType string, blockIDs []flowgo.Identifier) ([]flowgo.BlockEvents, error)
+	GetEventsForHeightRange(eventType string, startHeight, endHeight uint64) ([]flowgo.BlockEvents, error)
+
+	ExecuteScript(script []byte, arguments [][]byte) (*types.ScriptResult, error)
+	ExecuteScriptAtBlockHeight(script []byte, arguments [][]byte, blockHeight uint64) (*types.ScriptResult, error)
+	ExecuteScriptAtBlockID(script []byte, arguments [][]byte, id flowgo.Identifier) (*types.ScriptResult, error)
+
+	GetAccountStorage(address flowgo.Address) (*types.AccountStorage, error)
+
+	SendTransaction(tx *flowgo.TransactionBody) error
+	AddTransaction(tx flowgo.TransactionBody) error
+}
+
+type AutoMineCapable interface {
+	EnableAutoMine()
+	DisableAutoMine()
+}
+
+type ExecutionCapable interface {
+	ExecuteAndCommitBlock() (*flowgo.Block, []*types.TransactionResult, error)
+	ExecuteNextTransaction() (*types.TransactionResult, error)
+	ExecuteBlock() ([]*types.TransactionResult, error)
+	CommitBlock() (*flowgo.Block, error)
+}
+
+// Emulator defines the method set of an emulated blockchain.
+type Emulator interface {
```

Comment:
I ❤️ this embedding 
One questin, do you think it would make sense to name the package `emulator` instead of `blockchain`. I think when using the blockchain type implementation of this interface, the consumer would access it like: `blockchain.NewBlockchain` which is a bit vague, where with change it would be: `emulator.New()`. 
That also in any case says there is stuttering in the current name, if you decide to keep it same, change NewBlockchain to New.

Reply:
emulator.New() seems better for me too, just I didn't want to move it to root somehow. Not sure what is the best solution to use emulator.  Maybe I can expose this from root with some single file. What do you think?

Reply:
I made some moving around, I think would be ok.

Reply:
I think you should just rename the folder `blockcahain` to `emulator` and package will also then be renamed to emualtor.

Reply:
Doesn’t this cause cyclic import? Not sure if I tried like that but I tried something similar and hit with import cycle

Reply:
I wouldn't keep the root package emulator, since nothing is in it. The only emulator package would be the older blockchain package.

Reply:
If the only emulator package is this one then it shouldn't cause any cycic imports I believe

Reply:
@sideninja  ah you mean people using will import something like "github.com/onflow/flow-emulator/emulator" ?